Store schedule monitoring configurations in its own singleton

m-bymike opened the pull request:
Instead of declaring runtime properties in the macros, we delegate storage of configuration properties to an extra class.

This aims at issue #103.


I open this pull request with a potential solution to #103, and I'm very open to discuss my approach. It should serve as a conversation starter, I don't consider it done by now :).

m-bymike commented:
:)

freekmurze commented on src/Support/ScheduledTasks/ScheduleMonitoringConfigurationsRepository.php:
```diff
@@ -0,0 +1,78 @@
+
+class ScheduleMonitoringConfigurationsRepository
+{
+    private array $store = [];
```
Let's add a docblock that shows the content of this array.

Alternatively you could introduce a `MonitoredScheduledTask` class and put that in the array

freekmurze commented on src/Support/ScheduledTasks/ScheduleMonitoringConfigurationsRepository.php:
```diff
@@ -0,0 +1,78 @@
+    }
+
+
+    private function setProperty(object $target, string $key, mixed $value): void
```
We don't use `private` in our OSS code. Could you make everything that is now `private` `protected`?

freekmurze commented on src/ScheduleMonitorServiceProvider.php:
```diff
@@ -111,32 +110,40 @@ protected function registerEventHandlers(): self
-            $this->monitorName = $monitorName;
+        $this->app->singleton(
+            Support\ScheduledTasks\ScheduleMonitoringConfigurationsRepository::class,
+            static fn () => new Support\ScheduledTasks\ScheduleMonitoringConfigurationsRepository(),
```
Import this namespace

freekmurze replied:
Remove the `static` keyword

freekmurze commented on src/ScheduleMonitorServiceProvider.php:
```diff
@@ -111,32 +110,40 @@ protected function registerEventHandlers(): self
+        );
+
+        /** @var Support\ScheduledTasks\ScheduleMonitoringConfigurationsRepository $scheduleMonitoringConfigurationsRepository */
+        $scheduleMonitoringConfigurationsRepository = $this->app->make(Support\ScheduledTasks\ScheduleMonitoringConfigurationsRepository::class);
```
Rename this to the more human `monitoredScheduledTasks`

freekmurze commented on src/Support/Concerns/UsesScheduleMonitoringConfigurationsRepository.php:
```diff
@@ -0,0 +1,13 @@
+
+use Spatie\ScheduleMonitor\Support\ScheduledTasks\ScheduleMonitoringConfigurationsRepository;
+
+trait UsesScheduleMonitoringConfigurationsRepository
```
rename to `usesMonitoredScheduledTasks`

freekmurze commented on src/Support/ScheduledTasks/ScheduleMonitoringConfigurationsRepository.php:
```diff
@@ -0,0 +1,78 @@
+
+namespace Spatie\ScheduleMonitor\Support\ScheduledTasks;
+
+class ScheduleMonitoringConfigurationsRepository
```
Rename to `monitoredScheduledTasks`

freekmurze reviewed (changes requested):
Seems good. Did you already test these changes in a real Laravel app?

m-bymike commented:
Hi @freekmurze ,

thank you for the review, I addressed all suggested changes.

So far I tested the changes only locally (including the latest version) in a large project where it works as expected.

freekmurze commented on src/Support/Concerns/UsesMonitoredScheduledTasks.php:
```diff
@@ -0,0 +1,13 @@
+
+trait UsesMonitoredScheduledTasks
+{
+    public function getScheduleMonitoringConfigurationsRepository(): MonitoredScheduledTasks
```
Rename this to `getMonitoredScheduledTasks`

m-bymike replied:
Thank you, missed this one.

✅  Renamed

freekmurze commented:
Seems that the tests are failing, could you take a look at that too?

m-bymike commented:
> Seems that the tests are failing, could you take a look at that too?

Just did, one of them uncovered a line that I missed, which has now been fixed (534e4e07eec4d8a1702f4406ef263ef4620e624a). The other failing test was because of a difference in the name of the testing environment. I chose to use a dynamic way to retrieve the name of the current environment (00045a40ed70a51001e938ea4592a22efdc8f242). 

Then I figured that the tests only include PHP 8.0 - 8.2, so I added 8.3 and 8.4. At the same time I removed 8.0 as it is EoL (3811a44f90de4a176d02ee7e23896fed1903d028).

And I also updated the PHP version requirement in the `composer.json` file to `^8.1` (f0298c0324ae8f6e28c00b937306f8ef22d56b09).

freekmurze commented:
Thanks!
